File: src/typescript/helpers/sort/bubble.ts
type BubbleSortResult = {
  sorted: number[];
  comparisons: number;
  swaps: number;
};

export function bubbleSort(arr: number[]): BubbleSortResult {
  const result = [...arr];
  let comparisons = 0;
  let swaps = 0;
  let sorted = false;

  while (!sorted) {
    sorted = true;

    for (let i = 0; i < result.length - 1; i++) {
      comparisons++;

      if (result[i] > result[i + 1]) {
        [result[i], result[i + 1]] = [result[i + 1], result[i]];
        swaps++;
        sorted = false;
      }
    }
  }

  return {
    sorted: result,
    comparisons,
    swaps,
  };
}

File: src/go/helpers/sorts/bubble.go
package sorts

func BubbleSort(arr []int) ([]int, int, int) {
	result := make([]int, len(arr))
	copy(result, arr)
	comparisons := 0
	swaps := 0
	sorted := false

	for !sorted {
		sorted = true
		for i := 0; i < len(result)-1; i++ {
			comparisons++
			if result[i] > result[i+1] {
				result[i], result[i+1] = result[i+1], result[i]
				swaps++
				sorted = false
			}
		}
	}

	return result, comparisons, swaps
}
